src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -223,18 +224,39 @@ private String serializeBatch(final Partition.PartitionKey partitionKey, final l
                 token);
         final String cursorSerialized = getContext().getObjectMapper().writeValueAsString(cursor);
 
-        final StringBuilder builder = new StringBuilder()
-                .append("{\"cursor\":")
-                .append(cursorSerialized);
+        final CountingOutputStream countingOutputStream = new CountingOutputStream(outputStream) {
+            @Override
+            public void flush() throws IOException {
+                // block flushing
+            }
+        };
+        final Writer writer = new OutputStreamWriter(countingOutputStream);
+
+        writer.write("{\"cursor\":");
+        writer.write(cursorSerialized);
         if (!events.isEmpty()) {
-            builder.append(",\"events\":[");
-            events.forEach(event -> builder.append(event).append(","));
-            builder.deleteCharAt(builder.length() - 1).append("]");
+            writer.write(",\"events\":[");
+            writer.flush();
+            boolean first = true;
+            for (final byte[] event : events) {
+                if (!first) {
+                    countingOutputStream.write(',');
+                }
+                countingOutputStream.write(event);
+                first = false;
+            }
+            countingOutputStream.write(']');
         }
-        metadata.ifPresent(s -> builder.append(",\"info\":{\"debug\":\"").append(s).append("\"}"));
-
-        builder.append("}").append(EventStream.BATCH_SEPARATOR);
-        return builder.toString();
+        if (metadata.isPresent()) {
+            writer.write(",\"info\":{\"debug\":\"");
+            writer.write(metadata.get());
+            writer.write("\"}");
+            writer.flush();
+        }
+        writer.append("}").append(EventStream.BATCH_SEPARATOR);
```

Comment:
I guess that just slipped there. I guess it would be consistent to use `write`. I'll extract the stream writing logic from `EventStream` and `StreamingState` and make it consistent.